init.lua

```diff
@@ -1,77 +1,302 @@
 		end
-		timer:start(update_interval*3)
 	end
 end
 
-function wielded_light.update_light_by_item(item, pos)
-	local stack = ItemStack(item)
-	local light_level = shiny_items[stack:get_name()]
-	local itemdef = stack:get_definition()
-	if not light_level and not itemdef then
+local timer = 0
+local function global_timer_callback(dtime)
+
+	timer = timer + dtime;
+	if timer < update_interval then
 		return
 	end
+	timer = 0
 
-	light_level = light_level or ((itemdef.light_source or 0) - level_delta)
+	for _, player in pairs(minetest.get_connected_players()) do
+		for _,callback in pairs(update_player_callbacks) do
+			callback(player)
+		end
+	end
 
-	if light_level > 0 then
-		wielded_light.update_light(pos, light_level)
+	for _,callback in pairs(update_callbacks) do
+		callback()
 	end
+
+	-- Look at each tracked entity and update its position
+	for uid, entity in pairs(tracked_entities) do
+		if entity_still_exists(entity) then
+			update_entity(entity)
+		else
+			tracked_entities[uid] = nil
+		end
+	end
+
+	-- Recalculate light levels
+	for pos,_ in pairs(light_recalcs) do
+		recalc_light(pos)
+	end
+	light_recalcs = {}
+end
+
+function recalc_light(pos)
+	if not active_lights[pos] then return end
+
+	local any_light = false
+	local max_light = 0
+	for id, light_level in pairs(active_lights[pos]) do
+		any_light = true
+		if light_level > max_light then
+			max_light = light_level
+		end
+	end
+	local pos_vec = minetest.string_to_pos(pos)
+	if not any_light then
+		active_lights[pos] = nil
+		minetest.swap_node(pos_vec, { name = "air" })
+		return
+	end
+	if max_light == 0 then
+		minetest.swap_node(pos_vec, { name = "air" })
+		return
+	end
+
+	max_light = math.min(max_light, max_light_level)
+	local old_value = 0
+	local name = minetest.get_node(pos_vec).name
+	if wielded_light.is_wielded_light(name) then
+		old_value = wielded_light.level_of_wielded_light(name)
+	end
+	if old_value ~= max_light then
+		minetest.swap_node(pos_vec, {
+			name = wielded_light.wielded_light_of_level(max_light)
+		})
+		minetest.get_node_timer(pos_vec):start(cleanup_interval)
+	end
+end
+
+function add_light(pos, id, light_level)
+	if not active_lights[pos] then
+		active_lights[pos] = {}
+	end
+	if active_lights[pos][id] ~= light_level then
+		-- minetest.log("error", "add "..id.." "..pos.." "..tostring(light_level))
+		active_lights[pos][id] = light_level
+		light_recalcs[pos] = true
+	end
+end
+
+function remove_light(pos, id)
+	if not active_lights[pos] then return end
+	-- minetest.log("error", "rem "..id.." "..pos)
+	-- minetest.after(update_interval, function ()
+	active_lights[pos][id] = nil
+	light_recalcs[pos] = true
+	-- end)
+end
+
+--- Shining API ---
+wielded_light = {}
+
+function wielded_light.register_lightstep(callback)
+	table.insert(update_callbacks, callback)
+end
+
+function wielded_light.register_player_lightstep(callback)
+	table.insert(update_player_callbacks, callback)
+end
+
+function wielded_light.wielded_light_of_level(light_level)
```

Comment:
Needs naming improvement. `get_light_node_name` or similar.

Reply:
Naming it like this makes it clearer that it's the inverse of `level_of_wielded_light`